src/drivers/net/marvell/aqc1xx.c

```diff
@@ -0,0 +1,634 @@
+		      ATL_RX_DMA_DESC_BUF_SIZE);
+
+	/*filter global ctrl */
+	ctrl = ATL_RPF_CTRL1_BRC_EN | ATL_RPF_CTRL1_L2_PROMISC |
+		ATL_RPF_CTRL1_ACTION | ATL_RPF_CTRL1_BRC_TSH;
+	ATL_WRITE_REG(ctrl, ATL_RPF_CTRL1);
+
+	/* vlan promisc */
+	ATL_WRITE_REG(ATL_RPF_CTRL2_VLAN_PROMISC, ATL_RPF_CTRL2);
+	/* enable rpf2 */
+	ATL_WRITE_REG(ATL_RPF2_CTRL_EN, ATL_RPF2_CTRL);
+
+	/* RX Packet Buffer 0 Register 1 */
+	ATL_WRITE_REG(ATL_RPB0_CTRL1_SIZE, ATL_RPB0_CTRL1);
+
+	/*RX Packet Buffer 0 Register 2 */
+	ctrl = ATL_RPB0_CTRL2_LOW_TSH | ATL_RPB0_CTRL2_HIGH_TSH |
+		ATL_RPB0_CTRL2_FC_EN;
+	ATL_WRITE_REG(ctrl, ATL_RPB0_CTRL2);
+
+	/*RPB global ctrl*/
+	ctrl = ATL_READ_REG(ATL_RPB_CTRL);
+	ctrl |= (ATL_RPB_CTRL_EN | ATL_RPB_CTRL_FC);
+	ATL_WRITE_REG(ctrl, ATL_RPB_CTRL);
+
+	/*TX data path*/
+	/* enable tpo2 */
+	ATL_WRITE_REG(ATL_TPO2_EN, ATL_TPO2_CTRL);
+	/* tpb global ctrl *** */
+	ATL_WRITE_REG(ATL_TPB0_CTRL1_SIZE, ATL_TPB0_CTRL1);
+
+	ctrl = ATL_TPB0_CTRL2_LOW_TSH | ATL_TPB0_CTRL2_HIGH_TSH;
+	/* tpb global ctrl *** */
+	ATL_WRITE_REG(ctrl, ATL_TPB0_CTRL2);
+
+	ctrl = ATL_READ_REG(ATL_TPB_CTRL);
+	ctrl |= (ATL_TPB_CTRL_EN | ATL_TPB_CTRL_PAD_EN);
+	/* tpb global ctrl */
+	ATL_WRITE_REG(ctrl, ATL_TPB_CTRL);
+
+	/*Enable rings*/
+	ATL_WRITE_REG(ATL_READ_REG(ATL_RING_TX_CTRL) | ATL_RING_TX_CTRL_EN,
+		      ATL_RING_TX_CTRL);
+	ATL_WRITE_REG(ATL_READ_REG(ATL_RING_RX_CTRL) | ATL_RING_RX_CTRL_EN,
+		      ATL_RING_RX_CTRL);
+
+	atl_rx_ring_fill(nic);
+
+	nic->hw_ops->start(nic);
+	DBG("AQUANTIA: code 0()\n");
+
+	return 0;
+err_alloc:
+	atl_ring_free(&nic->tx_ring);
+	atl_ring_free(&nic->rx_ring);
+	DBG("AQUANTIA: code NOMEM()\n");
+	return -ENOMEM;
+}
+
+/**
+* Close network device
+*
+* @v netdev		Network device
+*/
+static void atl_close(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+
+	nic->hw_ops->stop(nic);
+	/* rpb global ctrl */
+	ATL_WRITE_REG(0x0, ATL_RPB_CTRL);
+
+	/* tgb global ctrl */
+	ATL_WRITE_REG(0x0, ATL_TPB_CTRL);
+
+	ATL_WRITE_REG(ATL_READ_REG(ATL_RING_TX_CTRL) | (~ATL_RING_TX_CTRL_EN),
+		      ATL_RING_TX_CTRL);
+	ATL_WRITE_REG(ATL_READ_REG(ATL_RING_RX_CTRL) | (~ATL_RING_RX_CTRL_EN),
+		      ATL_RING_RX_CTRL);
+
+	/* clear itr mask */
+	ATL_WRITE_REG(0x0, ATL_ITR_MSKS);
+
+	/* Reset the NIC */
+	nic->hw_ops->reset(nic);
+
+	atl_ring_free(&nic->tx_ring);
+	atl_ring_free(&nic->rx_ring);
+}
+
+/**
+* Transmit packet
+*
+* @v netdev		Network device
+* @v iobuf		I/O buffer
+* @ret rc		Return status code
+*/
+int atl_transmit(struct net_device *netdev, struct io_buffer *iobuf)
+{
+	struct atl_nic *nic = netdev->priv;
+	struct atl_desc_tx *tx;
+	physaddr_t address;
+	size_t len;
+
+	/* Get next transmit descriptor */
+	if (atl_ring_full(&nic->tx_ring)) {
+		DBG("AQUANTIA: %p out of transmit descriptors\n", nic);
+		return -ENOBUFS;
+	}
+	tx = (struct atl_desc_tx *)nic->tx_ring.ring + nic->tx_ring.sw_tail;
+
+	/* Populate transmit descriptor */
+	address = virt_to_bus(iobuf->data);
```

Comment:
Please use `iob_dma()` so that your driver will work on systems with an IOMMU enabled.